core/frontend/src/render/webgl/ScreenSpaceEffect.ts

```diff
@@ -0,0 +1,251 @@
+
+    // NB: compile() will throw with WebGL error log if compile/link fails.
+    if (CompileStatus.Success !== program.compile())
+      throw new Error(`Failed to produce shader program for screen-space effect "${this._name}"`);
+
+    const technique = new SingularTechnique(program);
+    const techniqueId = System.instance.techniques.addDynamicTechnique(technique, this._name);
+
+    const effect = new ScreenSpaceEffect(techniqueId, this._name, this._shiftsPixels, this.shouldApply);
+    System.instance.screenSpaceEffects.add(effect);
+  }
+}
+
+class ScreenSpaceEffect {
+  public readonly techniqueId: TechniqueId;
+  public readonly name: string;
+  private readonly _shouldApply?: ShouldApply;
+  private readonly _shiftsPixels: boolean;
+
+  public constructor(techniqueId: TechniqueId, name: string, shiftsPixels: boolean, shouldApply?: ShouldApply) {
+    this.techniqueId = techniqueId;
+    this.name = name;
+    this._shouldApply = shouldApply;
+    this._shiftsPixels = shiftsPixels;
+  }
+
+  public shouldApply(target: Target): boolean {
+    if (target.isReadPixelsInProgress) {
+      // Effects only apply during readPixels() if they move pixels around (we need to move pixels in the pick buffers correspondingly).
+      return this._shiftsPixels;
+    }
+
+    return undefined === this._shouldApply || this._shouldApply(target.screenSpaceEffectContext);
+  }
+}
+
+class ScreenSpaceGeometry extends ViewportQuadGeometry {
+  public setTechniqueId(id: TechniqueId) {
+    this._techniqueId = id;
+  }
+}
+
+/** @internal */
+export class ScreenSpaceEffects {
+  private readonly _effects = new Map<string, ScreenSpaceEffect>();
+  private readonly _effectGeometry: ScreenSpaceGeometry;
+  private readonly _copyGeometry: SingleTexturedViewportQuadGeometry;
+  private readonly _workingArray: ScreenSpaceEffect[] = [];
+
+  public constructor() {
+    // We will change the geometry's TechniqueId before applying each technique.
+    const effectGeometry = ScreenSpaceGeometry.create(TechniqueId.Invalid);
+    assert(effectGeometry instanceof ScreenSpaceGeometry);
+    this._effectGeometry = effectGeometry;
+
+    // NB: We'll replace the texture each time we draw.
+    const copyGeometry = SingleTexturedViewportQuadGeometry.createGeometry(System.instance.lineCodeTexture!.getHandle()!, TechniqueId.CopyColor);
+    assert(undefined !== copyGeometry);
+    this._copyGeometry = copyGeometry;
+  }
+
+  public dispose(): void {
+    dispose(this._effectGeometry);
+    dispose(this._copyGeometry);
+  }
+
+  public add(effect: ScreenSpaceEffect): void {
+    if (undefined !== this._effects.get(effect.name))
+      throw new Error(`Screen-space effect "${effect.name}" is already registered.`);
+
+    this._effects.set(effect.name, effect);
+  }
+
+  /** Return true if any effects should be applied to this Target. */
+  public shouldApply(target: Target): boolean {
+    return 0 < this.getApplicableEffects(target).length;
```

Comment:
maybe change to (more readable imo) `this.getApplicableEffects(target).length > 0`

Reply:
if not preferred (it doesn't seem to be consistent with most of our comparison operator usage anyway), resolve this

Reply:
Pushed this change.